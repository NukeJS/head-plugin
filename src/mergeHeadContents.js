import compareTags from './compareTags.js';

export default function mergeHeadContents(currentHead, newHead, { shouldPersist = () => false } = {}) {
	const themeActive = Boolean(document.querySelector('[data-swup-theme]'));
	const addTags = getTagsToAdd(currentHead.children, newHead.children, { themeActive });
	const removeTags = getTagsToRemove(currentHead.children, newHead.children);

	// Remove tags in reverse to keep indexes, keep persistant elements
	removeTags.reverse()
		.filter(({ el }) => shouldManageTag(el))
		.filter(({ el }) => !shouldPersist(el))
		.forEach(({ el }) => currentHead.removeChild(el));

	// Insert tag *after* previous version of itself to preserve JS variable scope and CSS cascaade
	addTags
		.filter(({ el }) => shouldManageTag(el))
		.forEach(({ el, index }) => {
			currentHead.insertBefore(el, currentHead.children[index + 1] || null);
		});

	return {
		removed: removeTags.map(({ el }) => el),
		added: addTags.map(({ el }) => el),
	};
};

function getTagsToRemove(currentTags, newTags) {
	const removeTags = [];

	for (let i = 0; i < currentTags.length; i++) {
		let foundAt = null;

		for (let j = 0; j < newTags.length; j++) {
			if (compareTags(currentTags[i], newTags[j])) {
				foundAt = j;
				break;
			}
		}

		if (foundAt == null && currentTags[i].getAttribute('data-swup-theme') === null) {
			removeTags.push({ el: currentTags[i] });
		}
	}

	return removeTags;
};

function getTagsToAdd(currentTags, newTags, { themeActive }) {
	const addTags = [];

	for (let i = 0; i < newTags.length; i++) {
		let foundAt = null;

		for (let j = 0; j < currentTags.length; j++) {
			if (compareTags(currentTags[j], newTags[i])) {
				foundAt = j;
				break;
			}
		}

		if (foundAt == null) {
			addTags.push({ el: newTags[i], index: themeActive ? i + 1 : i });
		}
	}

	return addTags;
};

function shouldManageTag(el) {
	return el.localName !== 'title'; // swup manages title itself
}
