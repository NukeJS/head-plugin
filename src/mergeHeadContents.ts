type ElementCollection = { el: Element; index?: number }[];

export default function mergeHeadContents(
	currentHead: HTMLHeadElement,
	newHead: HTMLHeadElement,
	{ shouldPersist = () => false }: { shouldPersist?: (el: Element) => boolean } = {}
) {
	const currentTags = Array.from(currentHead.children);
	const newChildren = Array.from(newHead.children);

	const addTags = getTagsToAdd(currentTags, newChildren);
	const removeTags = getTagsToRemove(currentTags, newChildren);

	// Remove tags in reverse to keep indexes, keep persistant elements
	removeTags
		.reverse()
		.filter(({ el }) => shouldManageTag(el))
		.filter(({ el }) => !shouldPersist(el))
		.forEach(({ el }) => currentHead.removeChild(el));

	// Insert tag *after* previous version of itself to preserve JS variable scope and CSS cascade
	addTags
		.filter(({ el }) => shouldManageTag(el))
		.forEach(({ el, index = 0 }) => {
			currentHead.insertBefore(el.cloneNode(true), currentHead.children[index + 1] || null);
		});

	return {
		removed: removeTags.map(({ el }) => el),
		added: addTags.map(({ el }) => el)
	};
}

function getTagsToRemove(currentEls: Element[], newEls: Element[]): ElementCollection {
	return currentEls.reduce((tags, el) => {
		const isAmongNew = newEls.some((newEl) => compareTags(el, newEl));
		if (!isAmongNew) {
			tags.push({ el });
		}
		return tags;
	}, [] as ElementCollection);
}

function getTagsToAdd(currentEls: Element[], newEls: Element[]): ElementCollection {
	return newEls.reduce((tags, el, index) => {
		const isAmongCurrent = currentEls.some((currentEl) => compareTags(el, currentEl));
		if (!isAmongCurrent) {
			tags.push({ el, index });
		}
		return tags;
	}, [] as ElementCollection);
}

function shouldManageTag(el: Element) {
	// Let swup manage the title tag
	if (el.localName === 'title') {
		return false;
	}
	// Leave swup theme styles untouched
	if (el.matches('[data-swup-theme]')) {
		return false;
	}
	return true;
}

function compareTags(oldTag: Element, newTag: Element) {
	return oldTag.outerHTML === newTag.outerHTML;
}
