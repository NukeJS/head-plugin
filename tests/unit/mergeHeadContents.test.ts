// FILEPATH: /Users/philipp/Projects/@swup/head-plugin/tests/unit/mergeHeadContents.test.ts

import { describe, expect, it } from 'vitest';
import mergeHeadContents from '../../src/mergeHeadContents.js';

const createHead = (contents: string = '') => {
    const html = `<head>${contents.trim()}</head>`;
    const { head } = new window.DOMParser().parseFromString(html, 'text/html');
    return head;
};

const mergeHeads = (current: string, incoming: string, options: Parameters<typeof mergeHeadContents>[2] = {}) => {
    const currentHead = createHead(current);
    const incomingHead = createHead(incoming);
    mergeHeadContents(currentHead, incomingHead, options);
    return currentHead;
};

describe('mergeHeadContents', () => {
    it('adds and removes tags from new head', () => {
        const head = mergeHeads(
            '<meta name="a" content="b">',
            '<meta name="b" content="c">'
        );
        expect(head.innerHTML).toMatchSnapshot();
    });

    it('keeps tags present in new head and current head', () => {
        const head = mergeHeads(
            '<link rel="next" content="/"><link rel="prev" content="/">',
            '<link rel="next" content="/">'
        );
        expect(head.innerHTML).toMatchSnapshot();
    });

    it('removes tags from current head not present in new head', () => {
        const head = mergeHeads(
            '<link rel="next" content="/"><link rel="prev" content="/">',
            '<link rel="next" content="/">'
        );
        expect(head.innerHTML).toMatchSnapshot();
    });

    it('keeps the right order of tags', () => {
        const head = mergeHeads(
            '<link rel="1"><link rel="2"><link rel="3"><link rel="4"><link rel="5">',
            '<link rel="5"><link rel="6"><link rel="2"><link rel="1">'
        );
        expect(head.innerHTML).toMatchSnapshot();
    });

    it('leaves title tags alone', () => {
        const head = mergeHeads(
            '<meta name="a" content="b">',
            '<meta name="a" content="b"><title>Title</title>'
        );
        expect(head.innerHTML).toMatchSnapshot();
    });

    it('does not remove persistent tags', () => {
        const head = mergeHeads(
            '<link rel="stylesheet"><link rel="prev">',
            '<meta name="a" content="b">',
            { shouldPersist: (el) => el.matches('[rel="stylesheet"]')}
        );
        expect(head.innerHTML).toMatchSnapshot();
    });

    it('returns the updated tags', () => {
        const head = createHead('<meta name="a"><meta name="b">');
        const incoming = createHead('<meta name="a"><meta name="c"><meta name="d">');
        const result = mergeHeadContents(head, incoming);
        expect(result.removed.length).toBe(1);
        expect(result.added.length).toBe(2);
        expect(result.removed[0].getAttribute('name')).toBe('b');
        expect(result.added[0].getAttribute('name')).toBe('c');
        expect(result.added[1].getAttribute('name')).toBe('d');
    });
});
