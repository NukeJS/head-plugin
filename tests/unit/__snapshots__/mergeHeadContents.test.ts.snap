// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mergeHeadContents > adds and removes tags from new head 1`] = `"<meta name="b" content="c">"`;

exports[`mergeHeadContents > does not remove persistent tags 1`] = `"<link rel="stylesheet"><meta name="a" content="b">"`;

exports[`mergeHeadContents > keeps tags present in new head and current head 1`] = `"<link rel="next" content="/">"`;

exports[`mergeHeadContents > keeps the right order of tags 1`] = `"<link rel="1"><link rel="2"><link rel="6"><link rel="5">"`;

exports[`mergeHeadContents > leaves title tags alone 1`] = `"<meta name="a" content="b">"`;

exports[`mergeHeadContents > removes tags from current head not present in new head 1`] = `"<link rel="next" content="/">"`;
